docs: improved docs around listener config (#5373)
confluentinc/ksql: docs/operate-and-deploy/installation/server-config/index.md

<path-to-confluent>/bin/ksql-server-start <path-to-confluent>/etc/ksqldb/ksql-server.properties
```

Configuring Listeners of a ksqlDB Cluster
------------------------------------------

Multiple hosts are required to scale ksqlDB processing power and to do that, they must
form a cluster.  ksqlDB requires all hosts of a cluster to use the same `ksql.service.id`.

```properties
bootstrap.servers=kafkaBroker1:9092,kafkaBroker2:9092
ksql.service.id=my_application_
```

Once formed, many operations can be run using the client APIs exposed on `listeners`.

In order to utilize pull queries and their high availability functionality, the nodes within the

> cluster must be able to communication with each other. KsqlDB supports setups with either a single

Reviewer: ```suggestion
cluster must be able to communicate with each other. ksqlDB supports setups with either a single
```

shared listener for both client and internal communication, or dual single-purpose listeners. The

> following describes how to configure listeners depending on the nature of your environment and

Reviewer: ```suggestion
following section describes how to configure listeners depending on the nature of your environment and
```

requirements.

### Single listener

#### Single routable listener

If you want to configure your listener with an IP address or hostname that is resolvable and
routable from within the cluster, you might do the following:

```properties
# Hostname that other nodes can resolve to an routable IP:
listeners=https://ksqlHost56:8088

# Or, routable IP address:
listeners=https://192.168.1.101:8088
```

> In this setup, the node will share the first url in the `listeners` config as its internal

Reviewer: ```suggestion
In this setup, the node shares the first URL in the `listeners` config as its internal
```

> endpoint, which other nodes will use for inter-node communication. Inter-node communication will

Reviewer: ```suggestion
endpoint, which other nodes use for inter-node communication. Inter-node communication
```

> use the same listener as client communication.

Reviewer: ```suggestion
uses the same listener as client communication.
```

#### Single non-routable listener

> It's common to setup a service using special hostnames such as `localhost` or wildcard addresses

Reviewer: ```suggestion
It's common to set up a service using special host names, like `localhost`, or wildcard addresses,
```

> such as `0.0.0.0` or `[::]`. These have special meanings and are not appropriate for

Reviewer: ```suggestion
like `0.0.0.0` or `[::]`. These special host names have special meanings and are not appropriate for
```

> inter-node communication, as they are not routable from other machines. Similarly, if your

Reviewer: ```suggestion
inter-node communication, because they're not routable from other machines. Similarly, if your
```

> network is setup such that the IP or hostname you bind isn't resolvable or routable.

Reviewer: ```suggestion
network is set up such that the IP or hostname you bind isn't resolvable or routable.
```

> If you choose to use a non-routable listener, you must set `ksql.advertised.listener`, specifying a

Reviewer: ```suggestion
If you choose to use a non-routable listener, you must set `ksql.advertised.listener` and specify a
```

> url that is externally accessible and which will resolve to an endpoint defined in `listeners`.

Reviewer: ```suggestion
URL that is externally accessible and which will resolve to an endpoint defined in `listeners`.
```

```properties
# Non-routable wildcard ip:
listeners=http://0.0.0.0:8088

> # External accessible name that will resolve to the IP of the machine:

Reviewer: ```suggestion
# External accessible name that resolves to the IP of the machine:
```

ksql.advertised.listener=http://host1.internal.example.com:8088
```

> In this setup, the node will share the url in the `ksql.advertised.listener` config as its

Reviewer: ```suggestion
In this setup, the node shares the URL in the `ksql.advertised.listener` config as its
```

> internal endpoint, which other nodes will use for inter-node communication. Inter-node

Reviewer: ```suggestion
internal endpoint, which other nodes use for inter-node communication. Inter-node
```

> communication will use the same listener as client communication.

Reviewer: ```suggestion
communication use the same listener as client communication.
```

### Dual listeners

You may choose to configure internal communication to use a different listener to client

> ommunication, which allows for port filtering rules to deny clients access the internal listener, or

Reviewer: ```suggestion
communication, which enables port filtering rules to deny clients access the internal listener, or
```

the use of a different network interface for internal communication, for security or QoS reasons.

This can be achieved by setting the `ksql.internal.listener` configuration to start a second

> listener that will exclusively used for inter-node communication.

Reviewer: ```suggestion
listener that is used exclusively for inter-node communication.
```

> If running dual listeners to improve security, you may also wish to enable

Reviewer: ```suggestion
If you're running dual listeners to improve security, you may also wish to enable
```

[authentication and other security measures](security.md).

#### Dual routable listeners

> Where the internal IP or hostname used in the `ksql.internal.listener` configuration is externally

Reviewer: ```suggestion
If the internal IP address or hostname used in the `ksql.internal.listener` configuration is externally
```

> resolvable and routable, only `ksql.internal.listener` needs to configured to set the internal

Reviewer: ```suggestion
resolvable and routable, you only need to configure `ksql.internal.listener` to set the internal
```

listener, for example:

```properties
# Client listener:
listeners=https://192.168.1.101:8088

# Inter-node listener on different NIC:
ksql.internal.listener=https://192.168.1.102:8088
```

!!! note
    Only the `ksql.internal.listener` needs to be resolvable and routable from servers running other

>     nodes in the cluster. The `listener` configuration can be non-resolvable and non-routable, as

Reviewer: ```suggestion
    nodes in the cluster. The `listener` configuration can be non-resolvable and non-routable, because
```

>     clients' can connect using whatever url you choose.

Reviewer: ```suggestion
    clients can connect using whatever URL you choose.
```

> In this setup, the node will share the url in the `ksql.internal.listener` config as its

Reviewer: ```suggestion
In this setup, the node shares the URL in the `ksql.internal.listener` config as its
```

> internal endpoint, which other nodes will use for inter-node communication. Inter-node

Reviewer: ```suggestion
internal endpoint, which other nodes use for inter-node communication. Inter-node
```

> communication will use a different listener to client communication.

Reviewer: ```suggestion
communication uses a different listener to client communication.
```

#### Dual non-routable listeners

> Where the internal IP or hostname used in the `ksql.internal.listener` configuration is not

Reviewer: ```suggestion
If the internal IP address or hostname used in the `ksql.internal.listener` configuration is not
```

externally resolvable and routable, for example where it uses `localhost` or wildcard IPs such as

> `0.0.0.0` or `[::]`, both `ksql.internal.listener` and `ksql.advertised.listener` need to configured

Reviewer: ```suggestion
`0.0.0.0` or `[::]`, you must configure both `ksql.internal.listener` and `ksql.advertised.listener`
```

to set the internal listener:

```properties
# Client listener:
listeners=https://0.0.0.0:8088

# Inter-node listener on wildcard address and different port:
# Note: port 8099 could be locked down using port forward or other network tools.
ksql.internal.listener=https://0.0.0.0:8099

> # Url that other nodes can resolve and use to route requests to this node:

Reviewer: ```suggestion
# URL that other nodes can resolve and use to route requests to this node:
```

ksql.advertised.listener=http://host1.internal.example.com:8099
```

!!! note
    Only the `ksql.advertised.listener` needs to be resolvable and routable from servers running
    other nodes in the cluster. The `listener` configuration can be non-resolvable and non-routable,

>     as clients' can connect using whatever url you choose, and `ksql.internal.listener` is only used

Reviewer: ```suggestion
    because clients can connect using whatever URL you choose, and `ksql.internal.listener` is only used
```

    to start the listener.

> In this setup, the node will share the url in the `ksql.advertised.listener` config as its

Reviewer: ```suggestion
In this setup, the node shares the url in the `ksql.advertised.listener` config as its
```

> internal endpoint, which other nodes will use for inter-node communication. Inter-node

Reviewer: ```suggestion
internal endpoint, which other nodes use for inter-node communication. Inter-node
```

> communication will use a different listener to client communication.

Reviewer: ```suggestion
communication uses a different listener to client communication.
```
